# Logistic Regression from Scratch

In this notebook, we implement Logistic Regression using:
- Sigmoid Activation Function
- Binary Cross Entropy (BCE) Loss
- Gradient Descent

The target variable is binary:
- 0 → Negative Class
- 1 → Positive Class

In [17]:
import numpy as np
import pandas as pd

np.random.seed(42)

# 500 samples, 2 features
X = np.random.randn(500, 2)

# True weights and bias
w = np.array([2, -1])
b = 1

# Linear combination
z = X @ w + b

# Sigmoid probabilities
p = 1 / (1 + np.exp(-z))

# Generate labels from probabilities
y = np.random.binomial(1, p)

# DataFrame
df = pd.DataFrame(X, columns=['feature1', 'feature2'])
df['target'] = y

print(df.head())
print("\nClass Distribution:")
print(df['target'].value_counts())

   feature1  feature2  target
0  0.496714 -0.138264       1
1  0.647689  1.523030       1
2 -0.234153 -0.234137       1
3  1.579213  0.767435       1
4 -0.469474  0.542560       0

Class Distribution:
target
1    315
0    185
Name: count, dtype: int64


## Step 2: Initialize Parameters

We initialize:
- Weights (m) for each feature
- Bias (b)

These parameters will be updated using Gradient Descent to minimize the Binary Cross Entropy (BCE) loss.

In [37]:
b = 0
m = np.array([1,1])

## Logistic Regression from Scratch

Logistic Regression is a classification algorithm used to predict the probability of a binary outcome.

Unlike Linear Regression, Logistic Regression applies the Sigmoid Activation Function to map predictions to probabilities between 0 and 1.

### Sigmoid Function

σ(z) = 1 / (1 + e^(-z))

where:

z = w₁x₁ + w₂x₂ + ... + wₙxₙ + b

### Training Process

1. Compute the linear combination (z)
2. Apply the sigmoid function to obtain probabilities
3. Calculate Binary Cross-Entropy (BCE) loss
4. Compute gradients with respect to weights and bias
5. Update parameters using Gradient Descent
6. Repeat until convergence

The objective is to minimize the BCE loss and learn the optimal decision boundary.

In [35]:
learning_rate = 0.1
epochs = 1000
for i in range(epochs):
    z = b + np.dot(m,X.T)
    y_hat = 1/(1+np.exp(-z))
    db = np.sum(y_hat-y)/X.shape[0]
    dm = np.dot((y_hat-y),X)/X.shape[0]
    b = b - (learning_rate)*db
    m = m - (learning_rate)*dm
print(b)
print(m)

1.028747009523433
[ 2.06134829 -1.27434207]


## Logistic Regression using Scikit-Learn

To validate the correctness of our implementation, we train a Logistic Regression model using Scikit-Learn.

Scikit-Learn provides an optimized implementation that automatically:
- Computes gradients
- Updates parameters
- Optimizes the loss function
- Generates predictions

We compare the learned:
- Weights (coefficients)
- Bias (intercept)

with those obtained from our from-scratch implementation.

Similar parameter values indicate that our implementation is mathematically correct.

In [36]:
from sklearn.linear_model import LogisticRegression
# Train model
lr = LogisticRegression()
lr.fit(X, y)

# Predictions
y_pred = lr.predict(X)

# Parameters
print("Bias (Intercept):", lr.intercept_[0])
print("Weights:", lr.coef_[0])


Bias (Intercept): 1.0039789034232713
Weights: [ 1.98085166 -1.22971051]


## Results

The Logistic Regression model was implemented from scratch using:

- Sigmoid activation
- Binary Cross Entropy Loss
- Gradient Descent

The learned parameters closely match those obtained using Scikit-Learn, validating the correctness of the implementation.

From Scratch:
- Bias ≈ 1.03
- Weights ≈ [2.06, -1.27]

Scikit-Learn:
- Bias ≈ 1.00
- Weights ≈ [1.98, -1.23]

The small differences arise from optimization and numerical precision.